### PR TITLE
Vue3:(feat) add source decorator vue template and setup script + supports of multi slots 

### PR DESCRIPTION
**Pull request by chakAs3**
Issue: #13917

## What I did
I have added a source decorator to renders/vue3 to show a pretty and practical source, multiple components are supported also, I have added source decorators.tests as well, to cover most cases.
image

source supports  slot default and named slot,  however, I checked the current render is not handling the slot in default Story, you still have to pass the template here, I may be working on this part also if you want @shilman  

## How to test

- [x] Is this testable with Jest or Chromatic screenshots? yes with jest and chromatic 
- [ ] Does this need a new example in the kitchen sink apps? no I don't thinko
- [ ] Does this need an update to the documentation? no does not

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Comment by chakAs3** [comment]
I have add a optional parameter `docs.source.withSetupScript` if set  true, the source generated with setup script 


by default if user does not set this params -->

image

if user sets this param 
image

source will be like this ->

image

**Comment by rmortes** [comment]
Absolute newbie here: Shouldn't the docs state `@click="onClick"` instead of `:onClick="onClick"`?

**Comment by chakAs3** [comment]
> 



> Absolute newbie here: Shouldn't the docs state `@click="onClick"` instead of `:onClick="onClick"`?

yes your are right, actually there is a bug when you set the argTypes on the story it overrides the table.category actually this should be included in the source, I fixed it for now and working on  fixing the argTypes issue

**Comment by ndelangen** [comment]
@chakAs3 please ping me when this is ready for review and green in CI, I'll happily take a look.

**Comment by chakAs3** [comment]
> @chakAs3 please ping me when this is ready for review and green in CI, I'll happily take a look.

Yes It is green 👍 now. actually, i detected 1 bug ` vue-docgen-api` was driving me mad, even though I upgraded to the latest one. it not able to parse properly the props from  vue component file with script setup, if you pass them as Reference to  to `defineProps(<Props>())`
for now, i put the props directly on cli templates, until we fix this issue ( I may do a PR, on their repos if I get time  )

File: code/renderers/vue3/src/docs/sourceDecorator.ts [hunk 1]
```diff
@@ -0,0 +1,275 @@
+/* eslint-disable no-underscore-dangle */
+import { addons, useEffect } from '@storybook/preview-api';
+import type { ArgTypes, Args, StoryContext, Renderer } from '@storybook/types';
+
+import { SourceType, SNIPPET_RENDERED } from '@storybook/docs-tools';
+
+import { format } from 'prettier';
```
Comment (ndelangen) [inline code/renderers/vue3/src/docs/sourceDecorator.ts:7]: I'm worried about the bundle-size implications of this. Isn't prettier like 1MB?

Comment (ndelangen) [reply code/renderers/vue3/src/docs/sourceDecorator.ts:7]: What happens if we do not format the code, and let the syntax-highlighter showing the code do this instead?
It has prettier build-in as well. But the problem might be it't not setup for vue code?

Comment (chakAs3) [reply code/renderers/vue3/src/docs/sourceDecorator.ts:7]: i know you would comment on this, it is more for indentention, my initial code was without that but it was not really pretty

Comment (chakAs3) [reply code/renderers/vue3/src/docs/sourceDecorator.ts:7]: i know you would comment on this, it is more for indentention, my initial code was without that but it was not really pretty

Comment (chakAs3) [reply code/renderers/vue3/src/docs/sourceDecorator.ts:7]: but it is already bundled  i guess, and i use it for parsing code as well to get the template from the StoryFn

Comment (chakAs3) [reply code/renderers/vue3/src/docs/sourceDecorator.ts:7]: I'm handling beautifully  complex nested components 
image

Comment (chakAs3) [reply code/renderers/vue3/src/docs/sourceDecorator.ts:7]: @shilman  has added prettier before 
image

Comment (chakAs3) [reply code/renderers/vue3/src/docs/sourceDecorator.ts:7]: i have new commit to support multiple and nested components with v-bind and dynamic and static args in the same story 

image

Story source is like this 

image

**Review by ndelangen — APPROVED** [review]

**Comment by chakAs3** [comment]
hi @ndelangen  @shilman.  I removed the 'prettier' package from my deps,  actually, I found out that SourceContainer handles the formatting pretty well, I had just to pass the preset to the handler function. and then kind of lazy load prettier from global deps to parse the vue template from the storyFn().

File: code/renderers/vue3/src/docs/sourceDecorator.ts (hunk 1 above)
Comment (ndelangen) [reply code/renderers/vue3/src/docs/sourceDecorator.ts]: 👍

**Comment by kasperpeulen** [comment]
@chakAs3 Awesome work!

**Comment by ndelangen** [comment]
@chakAs3 I was comparing dist output sizes, and noticed a huge difference in the dist of render/vue2 and renderer/vue.
Where renderer/vue was a lot bigger.

This is likely due to this line here:
https://github.com/storybookjs/storybook/blob/ee72ba54f8f4a4d3e6a889432dbf4dff1444a99a/code/renderers/vue3/src/docs/sourceDecorator.ts#L7-L8

and here:
https://github.com/storybookjs/storybook/blob/ee72ba54f8f4a4d3e6a889432dbf4dff1444a99a/code/renderers/vue3/src/docs/sourceDecorator.ts#L146

Can we change that code to not depend on prettier?

**Comment by ndelangen** [comment]
At the very least if we're not doing the above, we need to add prettier as a devDependency to the package.
The bundler is now bundling prettier is 'by accident' but this isn't clear from the package.json.

**Pull request merged**